src/kernels/attention_kernels.h

```diff
@@ -596,4 +648,135 @@ void crossAttnShardedHead(T1 *output, const T1 *query, const float *attnMask, in
+void crossAttnByHead(T *output, const T *query, const T *key, const T *value, int qHeadNum, int kvHeadNum, int headSize,
+        int oStride, int qStride, int kvStride, int batchSize, const int *inputSeqLens, const int *pastSeqLens,
+        bool causal, const float **alibiSlopes, const float **attnMask, const float scale, int threadNum,
+        const Lambda1 &getKHead, const Lambda2 &getVHead) {
+
+    int responsibleHeads = qHeadNum;
+    int groupNum = qHeadNum / kvHeadNum;
+
+    // To get row offset for each sample/sequence inside the batch, and prepare score buffer
+    int inputOffsets[batchSize];
+    size_t scoreSizePerThr = 0;
+    for (int i = 0; i < batchSize; ++i) {
+        scoreSizePerThr = std::max(scoreSizePerThr, (size_t)inputSeqLens[i] * (inputSeqLens[i] + pastSeqLens[i]));
+        inputOffsets[i] = (i > 0 ? inputOffsets[i - 1] + inputSeqLens[i] : 0);
+    }
+
+    scoreSizePerThr = ALIGNED_SIZE(scoreSizePerThr, 16);
+    size_t scoreSize = scoreSizePerThr * threadNum;
+    float *scoreBuf = (float *)SimpleMemPool::instance().getBuffer("scoreBuf", sizeof(float) * scoreSize);
+
+#pragma omp parallel for collapse(2)
+    for (int b = 0; b < batchSize; ++b) {
+        for (int i = 0; i < responsibleHeads; ++i) {
+            // Copy current key to cached keys (if needed)
+            int kvHdx = i / groupNum;
+            auto keyMatInfo = getKHead(b, kvHdx);
+            auto valueMat = getVHead(b, kvHdx);
+            bool bCopyCache = (i % groupNum == 0);
+
+            // Q * K
+            auto Q = query + inputOffsets[b] * qStride + i * headSize;
+            auto S = scoreBuf + omp_get_thread_num() * scoreSizePerThr;
+
+            const int queryLen = inputSeqLens[b];
+            const int keyLen = pastSeqLens[b] + inputSeqLens[b];
+
+            if (bCopyCache) {
+                int m = queryLen;
+                int n = keyLen;
+                int lda = qStride;
+                int ldc = keyLen;
+
+                // Copy to Key cache and compute Query * Key
+                auto src = key + inputOffsets[b] * kvStride + kvHdx * headSize;
+                storeKVCache(keyMatInfo, src, pastSeqLens[b], inputSeqLens[b], headSize, kvStride);
+
+                gemmQK(Q, keyMatInfo, S, m, n, headSize, lda, ldc);
+            } else {
+                // Note: when KV cache is not copied by me, then 2 times gemm to avoid synchronization
+                int m = queryLen;
+                int n = pastSeqLens[b];
+                int lda = qStride;
+                int ldc = keyLen;
+                gemmQK(Q, keyMatInfo, S, m, n, headSize, lda, ldc);
+
+                int ldb = kvStride;
+                auto B = key + inputOffsets[b] * kvStride + kvHdx * headSize;
+                small_gemm_transb(Q, B, S + n, m, inputSeqLens[b], headSize, lda, ldb, ldc);
+            }
+
+            // Softmax(Q * K)
+            if (causal && alibiSlopes == nullptr) {
```

Comment:
During the decoding phase, the softmax computing is the same whether causal is False（chatglm） or True, so “causal” can be omitted in the “if condition”

Reply:
Done.